packages/dds/tree/src/feature-libraries/treeTextCursor.ts

```diff
@@ -167,34 +141,180 @@ export class RootedTextCursor extends TextCursor {
+        // Skip top level, since root node in path is "undefined" and does not have a parent or index.
+        for (let height = 2; height < length; height += 2) {
+            const key = this.getStackedFieldKey(height - 1);
             path = {
                 parent: path,
-                parentIndex: this.indexStack[height],
-                parentField: this.keyStack[height],
+                parentIndex: this.getStackedNodeIndex(height),
+                parentField: key,
             };
         }
+
         path = {
             parent: path,
             parentIndex: this.index,
-            parentField: this.keyStack[length],
+            parentField: this.getStackedFieldKey(length - 1),
         };
         return path;
     }
+
+    public enterField(key: FieldKey): void {
+        // assert(this.mode === CursorLocationType.Nodes, "must be in nodes mode");
+        this.siblingStack.push(this.siblings);
+        this.indexStack.push(this.index);
+        this.index = -1;
+        this.siblings = key;
+    }
+
+    public get mode(): CursorLocationType {
+        return this.siblingStack.length % 2 === 0 ? CursorLocationType.Nodes : CursorLocationType.Fields;
+    }
+
+    public get pending(): boolean {
+        return false;
+    }
+
+    public nextField(skipPending: boolean): boolean {
+        if (this.mode === CursorLocationType.Nodes) {
+            // Implicitly enter first field
+            assert(this.index !== -1, "index should not be -1 in Nodes mode");
+            const fields = keys(this.getNode());
+            if (fields.length === 0) {
+                // No fields, so implicitly exit fields.
+                return false;
+            }
+
+            this.siblingStack.push(this.siblings);
+            this.siblings = fields;
+            this.indexStack.push(this.index);
+            this.index = 0;
+            return true;
+        }
+
+        // Already in "Fields" mode, so go to next field.
+        if (this.index === -1) {
+            // Navigated down to this field using a key, not iteration.
+            // Start iterating from this field:
+
+            // Iteration order is only defined for an individual iteration.
+            // If starting in the middle because accessed using `enterField`,
+            // we can unconditionally start iteration at the end.
+            this.upToNode();
+            return false;
+
+            // TODO: if requiring stable iteration order, use the below code:
+            // const key = this.siblings as FieldKey;
+            // const parents = this.siblingStack[this.siblingStack.length - 1] as readonly JsonableTree[];
+            // const parent: JsonableTree = parents[this.indexStack[this.indexStack.length - 1] as number];
+            // this.siblings = keys(parent);
+            // this.index = this.siblings.indexOf(key);
+            // // If key was not found, was in empty field, can start anywhere.
+            // this.index = 0;
+        }
```

Comment:
I'm supportive of punting cases we can't implement efficiently, but perhaps a 'fail()' would be a better choice?

Reply:
(BTW, I have to acknowledge the cleverness of your reasoning.)

Reply:
doing a fail here would need us to change the API and ban iterating fields in this case, maybe a third mode. Seems like it would add some API complexity and this workaround seems fine without incurring perf overhead. I think leaving it as is seems like the simplest option for now.